src/workerd/api/urlpattern-standard.h

```diff
@@ -0,0 +1,152 @@
+#include <kj/one-of.h>
+#include <kj/string.h>
+
+#include <optional>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace workerd::api {
+
+namespace urlpattern {
+#define URL_PATTERN_COMPONENTS(V)                                                                  \
+  V(Protocol, protocol)                                                                            \
+  V(Username, username)                                                                            \
+  V(Password, password)                                                                            \
+  V(Hostname, hostname)                                                                            \
+  V(Port, port)                                                                                    \
+  V(Pathname, pathname)                                                                            \
+  V(Search, search)                                                                                \
+  V(Hash, hash)
+
+// URLPattern is a Web Platform standard API for matching URLs against a
+// pattern syntax (think of it as a regular expression for URLs). It is
+// defined in https://wicg.github.io/urlpattern.
+// More information about the URL Pattern syntax can be found at
+// https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API
+class URLPattern final: public jsg::Object {
+ public:
+  class URLPatternRegexEngine {
+   public:
+    URLPatternRegexEngine() = default;
+    using regex_type = jsg::JsRef<jsg::JsRegExp>;
+    static std::optional<regex_type> create_instance(std::string_view pattern, bool ignore_case);
+    static std::optional<std::vector<std::optional<std::string>>> regex_search(
+        std::string_view input, const regex_type& pattern);
+    static bool regex_match(std::string_view input, const regex_type& pattern);
+  };
+
+  // A structure providing matching patterns for individual components
+  // of a URL. When a URLPattern is created, or when a URLPattern is
+  // used to match or test against a URL, the input can be given as
+  // either a string or a URLPatternInit struct. If a string is given,
+  // it will be parsed to create a URLPatternInit. The URLPatternInit
+  // API is defined as part of the URLPattern specification.
+  struct URLPatternInit {
+#define V(_, name) jsg::Optional<kj::String> name;
+    URL_PATTERN_COMPONENTS(V)
+#undef V
+    jsg::Optional<kj::String> baseURL;
+
+    JSG_STRUCT(protocol, username, password, hostname, port, pathname, search, hash, baseURL);
+
+    ada::url_pattern_init toAdaType() const;
+  };
+
+  // A struct providing the URLPattern matching results for a single
+  // URL component. The URLPatternComponentResult is only ever used
+  // as a member attribute of a URLPatternResult struct. The
+  // URLPatternComponentResult API is defined as part of the URLPattern
+  // specification.
+  struct URLPatternComponentResult final {
+    jsg::JsString input;
+    jsg::JsObject groups;
```

Comment:
Previously groups was well typed:
```
jsg::Dict<kj::String, kj::String> groups;
```
Why did we change that?

Reply:
Changing is fine, I think, might just need a type override tho.

Reply:
This actually made the code a lot more simpler and easier to implement. Also, I used a v8::Object::New constructor that takes size, key and values array to speed up the javascript object creation whereas the existing jsg::Dict doesn't use that constructor.

Reply:
Seems like we can amend jsg::Dict to use whatever api is best but I am indifferent to this if you think it's too much of a lift

Reply:
If you are comfortable with this: I can follow up after this PR lands?